### PR TITLE
cpu/stm32l0,l1: Fix ADC initialization order

### PR DESCRIPTION
**Pull request by crasbe**
### Contribution description

Most of this PR is already described in #20780.

The STM32L0 was enabled before the resolution bits were set, but any access to configuration registers is prohibited (and seemingly ignored by the microcontroller). Therefore, the resolution set in `adc_sample` was not actually applied and the resolution was always 12-bit.

The same issue was present in the STM32L1, however this ADC hardware implementation does not seem to ignore configuration register calls (even though they are prohibited when the ADC is on). However, the resolution bits were not masked in the configuration register before setting the new resolution, leading to the resolution being stuck at 6-bit (0b11). The resolution was only OR-ed to the register.

### Testing procedure

For convenience you can change the test application (`tests/periph/adc`) to print all resolution values at once, as described in https://github.com/RIOT-OS/RIOT/issues/20780#issuecomment-2482672894.

Due to the issue described in #21010, I recommend flashing the NUCLEO-L152RE board with the Mass Storage Driver of the ST-Link or doing a powercycle (unplug and plug the USB cable back in) after flashing. This issue is not related to these changes.

You can connect one of the A0 to A5 pins of the Nucleo to +3.3V and observe the output.
tl;dr: The output should look something like this, when the channel is maxed out on all implemented bit sizes. The two `-1` at the end show that the 14-bit and 16-bit resolution is not implemented.
```
expected:
ADC_LINE(5): 63 255 1023 4095 -1 -1

actual:
ADC_LINE(5): 4095 4095 4095 4095 -1 -1 (NUCLEO-L073RZ)
ADC_LINE(5): 63 63 63 63 -1 -1 (NUCLEO-L152RE)
```

<hr>

Nucleo-L073RZ Master:
```
main(): This is RIOT! (Version: 2025.01-devel-29-g76b9b)

RIOT ADC peripheral driver test

This test will sample all available ADC lines once every 100ms with
a 10-bit resolution and print the sampled results to STDIO


Successfully initialized ADC_LINE(0)
Successfully initialized ADC_LINE(1)
Successfully initialized ADC_LINE(2)
Successfully initialized ADC_LINE(3)
Successfully initialized ADC_LINE(4)
Successfully initialized ADC_LINE(5)
ADC_LINE(0): 73 32 14 6 -1 -1
ADC_LINE(1): 515 526 530 530 -1 -1
ADC_LINE(2): 325 152 74 39 -1 -1
ADC_LINE(3): 476 518 524 531 -1 -1
ADC_LINE(4): 482 514 537 536 -1 -1
ADC_LINE(5): 4095 4095 4095 4095 -1 -1
```

Nucleo-L073RZ this PR:
```
main(): This is RIOT! (Version: 2025.01-devel-31-g5de25c-pr/stm32l0_adc_fix)

RIOT ADC peripheral driver test

This test will sample all available ADC lines once every 100ms with
a 10-bit resolution and print the sampled results to STDIO


Successfully initialized ADC_LINE(0)
Successfully initialized ADC_LINE(1)
Successfully initialized ADC_LINE(2)
Successfully initialized ADC_LINE(3)
Successfully initialized ADC_LINE(4)
Successfully initialized ADC_LINE(5)
ADC_LINE(0): 1 2 4 8 -1 -1
ADC_LINE(1): 7 32 133 522 -1 -1
ADC_LINE(2): 5 10 19 52 -1 -1
ADC_LINE(3): 7 32 129 528 -1 -1
ADC_LINE(4): 7 32 132 537 -1 -1
ADC_LINE(5): 63 255 1023 4095 -1 -1
```

Nucleo-L152RE Master:
```
main(): This is RIOT! (Version: 2025.01-devel-29-g76b9b)

RIOT ADC peripheral driver test

This test will sample all available ADC lines once every 100ms with
a 10-bit resolution and print the sampled results to STDIO


Successfully initialized ADC_LINE(0)
Successfully initialized ADC_LINE(1)
Successfully initialized ADC_LINE(2)
Successfully initialized ADC_LINE(3)
Successfully initialized ADC_LINE(4)
Successfully initialized ADC_LINE(5)
ADC_LINE(0): 24 27 29 30 -1 -1
ADC_LINE(1): 25 29 30 31 -1 -1
ADC_LINE(2): 21 24 26 28 -1 -1
ADC_LINE(3): 23 26 28 29 -1 -1
ADC_LINE(4): 24 28 30 31 -1 -1
ADC_LINE(5): 63 63 63 63 -1 -1
```

Nucleo-L152RE with the fixes applied:
```
main(): This is RIOT! (Version: 2025.01-devel-31-g5de25c-pr/stm32l0_adc_fix)

RIOT ADC peripheral driver test

This test will sample all available ADC lines once every 100ms with
a 10-bit resolution and print the sampled results to STDIO


Successfully initialized ADC_LINE(0)
Successfully initialized ADC_LINE(1)
Successfully initialized ADC_LINE(2)
Successfully initialized ADC_LINE(3)
Successfully initialized ADC_LINE(4)
Successfully initialized ADC_LINE(5)
ADC_LINE(0): 23 110 473 1967 -1 -1
ADC_LINE(1): 28 123 503 2039 -1 -1
ADC_LINE(2): 21 98 423 1813 -1 -1
ADC_LINE(3): 22 103 447 1879 -1 -1
ADC_LINE(4): 22 111 481 1999 -1 -1
ADC_LINE(5): 63 255 1023 4095 -1 -1
```

### Issues/PRs references

This PR is a requirement for #20971, otherwise the changes in that PR can't be tested.

### CODE REVIEW

File: cpu/stm32/periph/adc_l1.c [hunk 1]
```diff
@@ -157,12 +155,18 @@ int32_t adc_sample(adc_t line, adc_res_t res)
...
-    /* set resolution, conversion channel and single read */
+    /* mask and set resolution, conversion channel and single read */
+    ADC1->CR1 &= ~ADC_CR1_RES_Msk;
     ADC1->CR1 |= res & ADC_CR1_RES;
     ADC1->SQR1 &= ~ADC_SQR1_L;
     ADC1->SQR5 = adc_config[line].chan;
 
-    /* wait for regulat channel to be ready*/
+    /* only set ADON when ADONS bit is cleared (ADC not ready) */
+    if(!(ADC1->SR & ADC_SR_ADONS_Msk)) {
```
Comment (benpicco) [inline cpu/stm32/periph/adc_l1.c:165]:
```suggestion
    if (!(ADC1->SR & ADC_SR_ADONS_Msk)) {
```

**Review by benpicco — COMMENTED** [review]
Looks goood to me!

**Comment by benpicco** [comment]
Please squash directly - I trust your testing.

**Review by benpicco — APPROVED** [review]

File: cpu/stm32/periph/adc_l1.c [hunk 2]
```diff
@@ -157,12 +155,18 @@ int32_t adc_sample(adc_t line, adc_res_t res)
     /* lock and power on the ADC device  */
     prep();
 
-    /* set resolution, conversion channel and single read */
+    /* mask and set resolution, conversion channel and single read */
+    ADC1->CR1 &= ~ADC_CR1_RES_Msk;
     ADC1->CR1 |= res & ADC_CR1_RES;
```
Comment (kfessel) [inline cpu/stm32/periph/adc_l1.c:160]: this hould result in 1 less read write
```suggestion
    /* mask and set resolution, conversion channel and single read */
    ADC1->CR1 = (res & ADC_CR1_RES) | ( ADC1->CR1 & ~ADC_CR1_RES_Msk);
```

Comment (crasbe) [reply cpu/stm32/periph/adc_l1.c:160]: I assumed that the compiler would optimize that anyways, but I'll add your suggestion.

My personal preference is to put the masking in front of everything, because then you can concatenate register modifications after that.

```suggestion
    /* mask and set resolution, conversion channel and single read */
    ADC1->CR1 = (ADC1->CR1 & ~ADC_CR1_RES_Msk) | (res & ADC_CR1_RES);
```

Comment (maribu) [reply cpu/stm32/periph/adc_l1.c:160]: I agree that readability would be better with separate statements. The issue is that when `volatile` variables are involved, every read and write operation is performed as specified.

```C
#include <stdint.h>

extern volatile uint32_t foo_reg;

void update_field_in_foo(uint8_t new_field_value)
{
    /* clear out old field value */
    foo_reg &= ~(0xff);
    /* write new field value */
    foo_reg |= new_field_value;
}

void update_field_in_foo_fast(uint8_t new_field_value)
{
    /* avoid multiple writes by using local variable for modification */
    uint32_t tmp = foo_reg;

    /* clear out old field value */
    tmp &= ~(0xff);
    /* update field new field value */
    tmp |= new_field_value;

    /* write back updated register */
    foo_reg = tmp;
}
```

Compiles to (with `arm-none-eabi-gcc` in 14.2.0 and `-Os`)

```asm
update_field_in_foo(unsigned char):
        ldr     r3, .L2
        ldr     r2, [r3]
        bic     r2, r2, #255
        str     r2, [r3]
        ldr     r2, [r3]
        orr     r0, r0, r2
        str     r0, [r3]
        bx      lr
.L2:
        .word   foo_reg
update_field_in_foo_fast(unsigned char):
        ldr     r2, .L5
        ldr     r3, [r2]
        bic     r3, r3, #255
        orr     r0, r0, r3
        str     r0, [r2]
        bx      lr
.L5:
        .word   foo_reg
```

What is particularly counter-intuitive to see (well, for me at least) is:

```asm
        str     r2, [r3]
        ldr     r2, [r3]
```

Where the value just written is read back. But that is required due to `volatile`, as the compiler must assume the value of `foo_reg` is updated externally (e.g. by hardware, because it is something like a timer counter or an SPI input register or whatever).

So: You can go for more explicit code (which I personally also consider more readable), if you first pass the value through a local variable that is not `volatile` and rely on the optimizer to work as expected for every operation on the local variable. Only the initial load from the memory reg to the local variable and the final write back are `volatile` memory operations and will not be optimized.

**Comment by kfessel** [comment]
@crasbe: if you provide test instruction it is nice if you tell what test you are taking about ( added that info in your message)

I usually just include the command-line  e.g.:  

```
<RIOT>/tests/periph/adc/ > BOARD=nucleo-l073rz make flash term 
```
other people use the `make  -D <dir>`

File: cpu/stm32/periph/adc_l1.c [hunk 3]
```diff
@@ -157,12 +155,17 @@ int32_t adc_sample(adc_t line, adc_res_t res)
     /* lock and power on the ADC device  */
     prep();
 
-    /* set resolution, conversion channel and single read */
-    ADC1->CR1 |= res & ADC_CR1_RES;
+    /* mask and set resolution, conversion channel and single read */
+    ADC1->CR1 = (ADC1->CR1 & ~ADC_CR1_RES_Msk) | (res & ADC_CR1_RES);
```
Comment (kfessel) [inline cpu/stm32/periph/adc_l1.c:159]: something seems wrong to me to my with this application of res values

```
ADC1->CR1 = (ADC1->CR1 & ~ADC_CR1_RES_Msk) | ((res << ADC_CR1_RES_Pos) & ADC_CR1_RES_Msk );
```
would be what i expect but on the other hand  the values in `adc_res_t` are shifted 
```
 typedef enum {
    ADC_RES_6BIT  = (0x3 << 3),     /**< ADC resolution: 6 bit */
    ADC_RES_8BIT  = (0x2 << 3),     /**< ADC resolution: 8 bit */
    ADC_RES_10BIT = (0x1 << 3),     /**< ADC resolution: 10 bit */
    ADC_RES_12BIT = (0x0 << 3),     /**< ADC resolution: 12 bit */
    ADC_RES_14BIT = (0xfe),         /**< not applicable */
    ADC_RES_16BIT = (0xff)          /**< not applicable */
} adc_res_t;
```

but not by the correct ammount:

```
cmsis/l1/Include/stm32l152xd.h
999:#define ADC_CR1_RES_Pos                      (24U)                             
1000:#define ADC_CR1_RES_Msk                      (0x3UL << ADC_CR1_RES_Pos)         /*!< 0x03000000 */
1001:#define ADC_CR1_RES                          ADC_CR1_RES_Msk                   /*!< ADC resolution */
1002:#define ADC_CR1_RES_0                        (0x1UL << ADC_CR1_RES_Pos)         /*!< 0x01000000 */
1003:#define ADC_CR1_RES_1                        (0x2UL << ADC_CR1_RES_Pos)         /*!< 0x02000000 */
```

**Review by kfessel — COMMENTED** [review]
I was a bit confused by `ADC_<something>` vs `ADC_<something>_Msk` seem like one is alias for the other e.g.:

```
cmsis/l1/Include/stm32l152xd.h
1000:#define ADC_CR1_RES_Msk                      (0x3UL << ADC_CR1_RES_Pos)         /*!< 0x03000000 */
1001:#define ADC_CR1_RES                          ADC_CR1_RES_Msk  
```

using only one ( I would prefer _Msk, i think these files before the pr prefer the one without ) would improve readability

File: cpu/stm32/periph/adc_l1.c (hunk 3 above)
Comment (kfessel) [reply cpu/stm32/periph/adc_l1.c:159]: would you please take another look at that

Comment (crasbe) [reply cpu/stm32/periph/adc_l1.c:159]: I think you looked into the header of the STM32L0, which has a different position of the resolution bits:

`cpu/stm32/include/periph/l0/periph_cpu.h` file: https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32/include/periph/l0/periph_cpu.h#L41-L51
```
#define HAVE_ADC_RES_T
typedef enum {
    ADC_RES_6BIT  = (0x3 << 3),     /**< ADC resolution: 6 bit */
    ADC_RES_8BIT  = (0x2 << 3),     /**< ADC resolution: 8 bit */
    ADC_RES_10BIT = (0x1 << 3),     /**< ADC resolution: 10 bit */
    ADC_RES_12BIT = (0x0 << 3),     /**< ADC resolution: 12 bit */
    ADC_RES_14BIT = (0xfe),         /**< not applicable */
    ADC_RES_16BIT = (0xff)          /**< not applicable */
} adc_res_t;
/** @} */
#endif /* ndef DOXYGEN */
```

The L1 already has the improved style for the resolution:

`cpu/stm32/include/periph/l1/periph_cpu.h` file: https://github.com/RIOT-OS/RIOT/blob/master/cpu/stm32/include/periph/l1/periph_cpu.h#L42-L56
```cpp
/**
 * @brief   Override the ADC resolution configuration
 * @{
 */
#define HAVE_ADC_RES_T
typedef enum {
    ADC_RES_6BIT  = (ADC_CR1_RES_0 | ADC_CR1_RES_1),    /**< ADC resolution: 6 bit */
    ADC_RES_8BIT  = (ADC_CR1_RES_1),                    /**< ADC resolution: 8 bit */
    ADC_RES_10BIT = (ADC_CR1_RES_0),                    /**< ADC resolution: 10 bit */
    ADC_RES_12BIT = (0x00),                             /**< ADC resolution: 12 bit */
    ADC_RES_14BIT = (0xfe),                             /**< not applicable */
    ADC_RES_16BIT = (0xff)                              /**< not applicable */
} adc_res_t;
/** @} */
#endif /* ndef DOXYGEN */
```

The resolution flag position (and registers) vary wildly between different STM32 families, and the implementation style in RIOT varies just as much, which was the motivation for #20971.

Comment (crasbe) [reply cpu/stm32/periph/adc_l1.c:159]: These are the relevant pages in the reference manuals. The register positions for both variants are correct.

https://www.st.com/resource/en/reference_manual/rm0367-ultralowpower-stm32l0x3-advanced-armbased-32bit-mcus-stmicroelectronics.pdf page 341

https://www.st.com/resource/en/reference_manual/cd00240193-stm32l100xx-stm32l151xx-stm32l152xx-and-stm32l162xx-advanced-arm-based-32-bit-mcus-stmicroelectronics.pdf page 291

File: cpu/stm32/periph/adc_l1.c [hunk 4]
```diff
@@ -157,12 +155,17 @@ int32_t adc_sample(adc_t line, adc_res_t res)
...
-    ADC1->CR1 |= res & ADC_CR1_RES;
+    /* mask and set resolution, conversion channel and single read */
+    ADC1->CR1 = (ADC1->CR1 & ~ADC_CR1_RES_Msk) | (res & ADC_CR1_RES_Msk);
     ADC1->SQR1 &= ~ADC_SQR1_L;
     ADC1->SQR5 = adc_config[line].chan;
 
-    /* wait for regulat channel to be ready*/
+    /* only set ADON when ADONS bit is cleared (ADC not ready) */
+    if (!(ADC1->SR & ADC_SR_ADONS_Msk)) {
+        ADC1->CR2 |= ADC_CR2_ADON;
```
Comment (kfessel) [inline cpu/stm32/periph/adc_l1.c:165]: _Msk and no _Msk

File: cpu/stm32/periph/adc_l1.c [hunk 5]
```diff
@@ -171,6 +174,10 @@ int32_t adc_sample(adc_t line, adc_res_t res)
     sample = (int)ADC1->DR;
     ADC1 -> SR &= ~ADC_SR_STRT;
 
+    /* wait for ADC to become ready before disabling it */
+    while (!(ADC1->SR & ADC_SR_ADONS)) {}
+    ADC1->CR2 &= ~ADC_CR2_ADON_Msk;
```
Comment (kfessel) [inline cpu/stm32/periph/adc_l1.c:179]: _Msk and no _Msk

**Review by kfessel — COMMENTED** [review]
the two marked snipetes use msk an no msk in opposite roles 

testing and (un)setting

**Comment by crasbe** [comment]
I just looked at the other code in the file and it does not use `_Msk` at all, only the code I introduced.
For consistency it would probably be best to not use the `_Msk` definitions at all then.

Likewise for the additions in my other PR #20971. Only the code introduced by me uses the `_Msk` definition.

**Comment by kfessel** [comment]
> I just looked at the other code in the file and it does not use `_Msk` at all, only the code I introduced. For consistency it would probably be best to not use the `_Msk` definitions at all then.
> 
> Likewise for the additions in my other PR #20971. Only the code introduced by me uses the `_Msk` definition.

I think either is ok (I would just prefer that there is one), diverting from cmsis isn't a problem as long a the manufacturer keeps doing that (microchip just removed their extra bitfield structs (which i also could not find in the cmsis manual) ) 

to me the _Msk variant is a little more telling what it actually is.

**Comment by crasbe** [comment]
@kfessel Can I resolve the conversations and squash the fixups?

**Comment by kfessel** [comment]
> @kfessel Can I resolve the conversations and squash the fixups?

yes

**Comment by maribu** [comment]
Thanks for this! I'll give Ben and Karl a bit of time to react before hitting merge.

**Comment by crasbe** [comment]
Thanks everyone for working with me on this. Even though it was a small change, I still learned a lot :)
